### PR TITLE
add indices to file validation

### LINKED ISSUE #4227: [BUG] CSV Uploader errors not surfacing relevant row indices
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When I submit a CSV with an invalid field on one or more rows, the error messages are surfaced without the relevant row information.
This is expected when the scope is on a report-level, like if there's a missing column, but not if the error is just isolated to one row.
Screen Shot 2022-09-01 at 1 46 30 PM
In this screenshot, "Row(s)" should be followed by the row numbers.

### Steps To Reproduce

Upload a file with an item-level error. An example would be to download the example file on the Upload spreadsheet page and then remove the data in the last name column for one or more of the rows.

### Expected Behavior

I expect to see row numbers for item-scope errors.

### Additional context

This is because our mutation is not getting the indices back in the errors/warnings we receive from RS: https://github.com/CDCgov/prime-simplereport/blob/468f7416eeadde228251ecf74b17388d3c86f2e1/frontend/src/app/testResults/uploads/operations.graphql

### Does this issue have a priority label?

- [X] I have added a priority label

### Has issue been assigned to a team?

- [X] I have added a team label

### PR DESCRIPTION
**Pull request by BobanL**
# BACKEND PULL REQUEST

## Related Issue

- Indices are not being returned in the error message for patient bulk upload and test result uploader
- Resolves #4227 

## Changes Proposed

- Indices get returned with the error messages
- A generic file validator has been created to support file validation across any potential csv we may need in the future. 
- File validator now will merge duplicate error messages and add the indices to the FeedbackMessage.
- Change Feedbackmessage indices type to list to make it easier to work with.
- New file row interface was created as the base for both the PatientUploadRow and TestResultRow

## Additional Information
- Deployed in [dev5](https://dev5.simplereport.gov/)!

## Testing

- Patient Bulk Upload - upload with a valid file. Example file: [patient_valid.csv](https://github.com/CDCgov/prime-simplereport/files/10061005/patient_valid.csv)
- Patient Bulk Upload - upload a file with multiple errors on different lines and validate the response message.
Example file: [patient_errors.csv](https://github.com/CDCgov/prime-simplereport/files/10061007/patient_errors.csv)
- Test Results Upload - upload a valid file. Example file: [test-results-upload-valid.csv](https://github.com/CDCgov/prime-simplereport/files/10061008/test-results-upload-valid.csv)
- Test Results Upload - upload a file with multiple errors and validate the response message. Example file: [test-results-upload-invalid-values.csv](https://github.com/CDCgov/prime-simplereport/files/10061010/test-results-upload-invalid-values.csv)

### CODE REVIEW

**Comment by emmastephenson** [comment]
Mostly this LGTM, but we should probably sync with Better Data on a unified way to parse/send these index errors - @DanielSass just did a similiar thing on their side: https://github.com/CDCgov/prime-simplereport/pull/4688

File: backend/src/main/java/gov/cdc/usds/simplereport/validators/PatientBulkUploadFileValidator.java
Comment (georgehager) [inline backend/src/main/java/gov/cdc/usds/simplereport/validators/PatientBulkUploadFileValidator.java:87]: Why do we need to redefine `currentRow` in a new var? I don't see it being modified before incrementing 🤔

Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/validators/PatientBulkUploadFileValidator.java:87]: This is a recommendation by intellij to not use non final variables in lambdas. [Read more here](https://stackoverflow.com/questions/34865383/variable-used-in-lambda-expression-should-be-final-or-effectively-final)
image

**Comment by georgehager** [comment]
Works nicely, and shows me multiple errors as expected! I'm gonna hold off on doing a review since feedback from BetterData is needed 🙂

**Comment by BobanL** [comment]
> Mostly this LGTM, but we should probably sync with Better Data on a unified way to parse/send these index errors - @DanielSass just did a similiar thing on their side: #4688

> Works nicely, and shows me multiple errors as expected! I'm gonna hold off on doing a review since feedback from BetterData is needed 🙂

So it seems that #4688 was specific to adding error messages and row indices for rows that failed processing, but not for validating. I have confirmed that with @zdeveloper and will be working on also applying this change to the `TestResultFileValidator` to resolve ticket #4227.

File: backend/src/test/java/gov/cdc/usds/simplereport/validators/FileValidatorTest.java
Comment (emmastephenson) [inline backend/src/test/java/gov/cdc/usds/simplereport/validators/FileValidatorTest.java:196]: nit: can you specify in the test name that this is for results instead of patients?

File: backend/src/test/java/gov/cdc/usds/simplereport/validators/PatientUploadRowTest.java
Comment (emmastephenson) [inline backend/src/test/java/gov/cdc/usds/simplereport/validators/PatientUploadRowTest.java:13]: nice, thanks for the additional test coverage here!

File: backend/src/main/java/gov/cdc/usds/simplereport/config/FileValidatorConfiguration.java
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/config/FileValidatorConfiguration.java:9]: This is neat, I haven't seen this configuration style before! I'm curious, what's the usage if we're manually declaring `FileValidator<type>` where it's being used?

**Review by emmastephenson — DISMISSED** [review]
LGTM, loving the approach you took with a parent `FileValidator` class! 🙌

**Review by emyl3 — APPROVED** [review]
Tested both the results bulk upload and patient bulk upload on dev5! Thanks for fixing this up!!

File: backend/src/main/java/gov/cdc/usds/simplereport/config/FileValidatorConfiguration.java
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/FileValidatorConfiguration.java:9]: I'm not sure if this answers your question, but these are actually being used to autowire the validator in the services.  The only place we are declaring file validator is in the tests.
image
image

If the question was around why we are passing around `TestResultRow::new` into the functions, it's because java can't instantiate generic types for example in FileValidator, you cannot do `var patientUploadRow = new T()`. After doing a little more research it seems like another way to do accomplish creating the generic is by accepting a class into the constructor and then creating it with that class reference, but that way requires a lot of exception handling. I do think this way may look a little nicer passing in a class, but I'd prefer not having to worry about any error handling. 
image
image
image

References: https://stackoverflow.com/questions/2434041/instantiating-generics-type-in-java

File: backend/src/main/java/gov/cdc/usds/simplereport/validators/FileRow.java
Comment (georgehager) [inline backend/src/main/java/gov/cdc/usds/simplereport/validators/FileRow.java:6]: nit: the naming of this is a little confusing, could it be something like `FileRowValidator`? same for the implementation files, too. It makes sense in the project structure, I just found skimming through I had to think twice about the name.

File: backend/src/main/java/gov/cdc/usds/simplereport/config/FileValidatorConfiguration.java
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/config/FileValidatorConfiguration.java:9]: Ahhhh right, the `Autowire` use case makes a lot of sense. That's mostly what I was asking about - why it had to be a config bean vs. just another POJO. 

I agree with you on passing the class vs new - I wouldn't want to do the exception handling either, and I think the `TestResultRow::new` is still perfectly readable.

Thanks for the detailed explanation!

**Review by georgehager — APPROVED** [review]
looks great! 🌮  also tested locally and works as expected. I like the move to the generic FIleValidator!

**Review by emmastephenson — DISMISSED** [review]
Everything LGTM! I would still request that you await review from one of @mehansen @DanielSass or @nathancrtr though, since this has impacts on the TestResultUpload service. I know Merethe in particular had strong feelings about the row indices 😄

File: backend/src/test/java/gov/cdc/usds/simplereport/validators/PatientUploadRowTest.java
Comment (emmastephenson) [inline backend/src/test/java/gov/cdc/usds/simplereport/validators/PatientUploadRowTest.java:47]: Sonar is complaining about these `public` modifiers - do you mind removing them for the tests?

File: backend/src/main/java/gov/cdc/usds/simplereport/validators/FileRow.java
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/validators/FileRow.java:6]: The methods the interface defines right now are used for validation, but thats not necessarily what the class needs to be used for as seen in the [PatientBulkUploadService](https://github.com/CDCgov/prime-simplereport/blob/main/backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java#L104).  I did notice that it was in the validators folder which may create confusion so I will move it to the model folder instead!

Comment (georgehager) [reply backend/src/main/java/gov/cdc/usds/simplereport/validators/FileRow.java]: Awesome! Yes, I think moving it out of the validator will clear up the confusion! 🌮

**Review by mehansen — COMMENTED** [review]
partial review because I wasn't able to take a look at the tests yet but only have a couple minor comments so far

File: backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/FileRow.java
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/FileRow.java:7]: I'll change it! There are some items in that section that aren't even marked as required, [like comment](https://github.com/CDCgov/prime-simplereport/blob/f74cd4a9ddc93a5d081f84cc3cc511c8a70571ab/backend/src/main/java/gov/cdc/usds/simplereport/validators/TestResultFileValidator.java#L124-L125), and then [still being validated.](https://github.com/CDCgov/prime-simplereport/blob/f74cd4a9ddc93a5d081f84cc3cc511c8a70571ab/backend/src/main/java/gov/cdc/usds/simplereport/validators/TestResultFileValidator.java#L188-L189). Would it be OK to use the fields that are marked `required=true` only in that section?

File: backend/src/main/java/gov/cdc/usds/simplereport/validators/FileValidator.java
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/validators/FileValidator.java:59]: Thanks for explaining that to me! I wasn't really sure what the difference was between those two, but so does that means that that even the [csv error thing](https://github.com/CDCgov/prime-simplereport/pull/4730/files/6969cdc13b6e80df3c06fc913ed9da13705a40ac#diff-2b537a592254778dd76f364d15df290173d0d8fda42a9db279bde9f26f9f911dR43-R48) should also be `ITEM_SCOPE`? If that happens no where in the code are we using `REPORT_SCOPE` so it might be easy to delete that field from `ValueOrError`.

Comment (mehansen) [reply backend/src/main/java/gov/cdc/usds/simplereport/validators/FileValidator.java:59]: oh yup, you're right I think that should be ITEM_SCOPE too
I think if we had a validation that checked the header line for missing required fields (this is how RS does it) then that would be REPORT_SCOPE, but the way we have things written now it looks like we don't need that value at all

File: backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/FileRow.java
Comment (mehansen) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/FileRow.java]: this is a good point. `field.getPossibleError()` is only going to return something if `field` was created with `required=true` so yeah checking the others is kind of pointless and I do think it would be OK to only use required fields for that section. I could see it being potentially easier to upkeep if all the fields are checked because someone might make a field required or create a new one and forget to add it to the validate required fields method but now that you've renamed the method that should make it pretty clear imo

**Review by mehansen — COMMENTED** [review]
finished reviewing
looks great, thanks for doing this!
the only thing I see that would be good to change is the scope values. whatever you decide to do wrt validating required fields makes sense to me

File: backend/src/main/java/gov/cdc/usds/simplereport/validators/FileValidator.java
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/validators/FileValidator.java:59]: done in [77f7cf4](https://github.com/CDCgov/prime-simplereport/pull/4730/commits/77f7cf4310ed9e54caf780fdc0524c8c4b69ad2b)/[9d118b2](https://github.com/CDCgov/prime-simplereport/pull/4730/commits/9d118b2f0511931d7ca497f6a098c5ea906d140a)

File: backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/FileRow.java
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/FileRow.java]: I took a stab at something you said about someone forgetting to validate the field if it is required. This honestly may have introduced way more complexity than needed though because of reflection, [2f3e8a5](https://github.com/CDCgov/prime-simplereport/pull/4730/commits/2f3e8a5c689f7cafcbe11e7fb2c58e1a40a7a394).

File: backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/FileRow.java
Comment (mehansen) [inline backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/FileRow.java:46]: since we've checked the type, is it safe to cast what's inside the Field so that we can invoke the method directly?

```suggestion
    if (field.getType().equals(ValueOrError.class)) {
      ValueOrError valueOrError = (ValueOrError) field.get(fr);
      if (valueOrError != null) {
        return valueOrError.getPossibleError();
      }
```

File: backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/PatientUploadRow.java
Comment (mehansen) [inline backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/PatientUploadRow.java:75]: could we use `this.getClass().getDeclaredFields()` and move this up to `FileRow` as well?

**Review by mehansen — COMMENTED** [review]
I think this is cool, and gives us just one method to change in the validation code when adding another column on a row (two methods if we want to validate values in that new column)

File: backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/FileRow.java
Comment (BobanL) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/FileRow.java]: That's a great catch! I just updated it in b25fa8f

**Review by mehansen — APPROVED** [review]
🎉

**Review by georgehager — APPROVED** [review]
thank you! 🎉

**Pull request merged**
